[Wallet] get rid of pwalletMain, add simple CWallets infrastructure

jonasschnelli opened the pull request:
This is a very simple pull requests that removes the global `pwalletMain` pointer and replaces the calls with `CWallets::defaultWallet()`.

This PR leverages the internal global single wallet container to a vector. We have similar functionality on the Qt layer.

This will allow a step-by-step transition towards multiwallet support like proposed by @luke-jr in #8694.

jonasschnelli committed:
[Wallet] get rid of pwalletMain, add simple CWallets infrastructure

laanwj commented on src/init.cpp:
```diff
@@ -231,8 +230,8 @@ void Shutdown()
 #ifdef ENABLE_WALLET
-    if (pwalletMain)
-        pwalletMain->Flush(true);
+    if (CWallets::defaultWallet())
```
CWallets::FlushAllWallets() ?

jonasschnelli replied:
Yes. Changed it to `flushAllWallets(bool shutdown);`

paveljanik commented on src/wallet/wallet.h:
```diff
@@ -986,4 +984,37 @@ class CAccount
+    {
+        LOCK(cs_wallets);
+        wallets.push_back(newWallet);
+        int i = 0;
```
`i` looks unused here 8)

jonasschnelli replied:
That's black magic! :-)

Will remove it, had it there for debug purposes.

jonasschnelli commented on src/init.cpp:
```diff
@@ -1535,9 +1531,9 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
+    if (CWallets::defaultWallet()) {
         // Run a thread to flush wallet periodically
-        threadGroup.create_thread(boost::bind(&ThreadFlushWalletDB, boost::ref(pwalletMain->strWalletFile)));
+        threadGroup.create_thread(boost::bind(&ThreadFlushWalletDB, boost::ref(CWallets::defaultWallet()->strWalletFile)));
```
We keep the `ThreadFlushWalletDB` with "capacity" for a single wallet file for now.

MarcoFalke replied:
Why?

jonasschnelli replied:
1. Scope creeping
2. Moving target (`ThreadFlushWalletDB` is currently getting touches in different PRs).

MarcoFalke replied:
Fine, can be done later. Though, I think in case it is just a simple for loop, it should be part of this pull.

jonasschnelli replied:
I think we don't want a flush wallet thread per wallet. IMO it should be a single flush thread that loops though the wallet, check last-flush-time and optionally flushes.

MarcoFalke replied:
Jup, @luke-jr did this. (Create a loop in `ThreadFlushWalletDB` over all wallets)

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -986,4 +984,36 @@ class CAccount
+class CWallets
+{
+public:
+    static std::vector<CWallet *> wallets;
```
Why public?

jonasschnelli replied:
Yes. I had a `Namespace` at the beginning and though that there is no big need for a Class or even private static class variables. But yes, why not private. Will update.

paveljanik commented on src/qt/signverifymessagedialog.cpp:
```diff
@@ -142,7 +142,7 @@ void SignVerifyMessageDialog::on_signMessageButton_SM_clicked()
 
     CKey key;
-    if (!pwalletMain->GetKey(keyID, key))
+    if (!CWallets::defaultWallet()->GetKey(keyID, key))
```
There is no `if (pwalletMain)` test here in the original code?

MarcoFalke replied:
@paveljanik The dialog does not exist when there is no wallet.

paveljanik commented on src/wallet/rpcdump.cpp:
```diff
@@ -270,7 +270,7 @@ UniValue importprunedfunds(const UniValue& params, bool fHelp)
         throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
     uint256 hashTx = tx.GetHash();
-    CWalletTx wtx(pwalletMain,tx);
+    CWalletTx wtx(CWallets::defaultWallet(),tx);
```
SPC after `,` :-)

MarcoFalke replied:
Or just `git diff HEAD~ -U0|cfd`.

```sh
alias cfd='~/workspace/bitcoin/contrib/devtools/clang-format-diff.py -p1 -i -v'

paveljanik commented:
I'm all for supporting multiple wallets, so concept ACK.

But isn't this only the part of @luke-jr 's work (64bcde7c5535cda12fc20ded718558ca74b28191)? Can you please coordinate the work to save your time? ;-)

jonasschnelli commented:
I really like @luke-jr https://github.com/bitcoin/bitcoin/pull/8694, though, my experience with lager pull requests tells me, that #8694 will have a hard time getting in as a whole. I try to slice out independent chunks, improve them and PR them separately. This is how I get you all reviewing parts of #8694.

jonasschnelli commented:
Fixed nits.

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -986,4 +984,37 @@ class CAccount
+    static void destruct()
+    {
+        LOCK(cs_wallets);
+        for(CWallet *pWallet : wallets)
```
μNit

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 656d344..d690e48 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -61 +61 @@ const uint256 CMerkleTx::ABANDON_HASH(uint256S("00000000000000000000000000000000
-std::vector<CWallet *> CWallets::wallets = {};
+std::vector<CWallet*> CWallets::wallets = {};
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index 390210d..4fb69dd 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -991 +991,2 @@ private:
-    static std::vector<CWallet *> wallets;
+    static std::vector<CWallet*> wallets;
+
@@ -1000 +1001 @@ public:
-    static void addWallet(CWallet *newWallet)
+    static void addWallet(CWallet* newWallet)
@@ -1008 +1009 @@ public:
-        for(CWallet *pWallet : wallets)
+        for (CWallet* pWallet : wallets)
@@ -1014 +1015 @@ public:
-        for(CWallet *pWallet : wallets)
+        for (CWallet* pWallet : wallets)

maflcko reviewed (approved):
utACK 9cc96b59f4bcf708ae370679ef41c6f8f2af3a14

luke-jr commented on src/wallet/wallet.h:
```diff
@@ -986,4 +984,37 @@ class CAccount
+    {
+        LOCK(cs_wallets);
+        for(CWallet *pWallet : wallets)
+            pWallet->Flush(shutdown);
```
Does cs_wallets need to be held for access to (flushing) the wallet itself, or only for the wallets vector? If the former, then defaultWallet will fail to guard it; if the latter, the lock here is unnecessarily long-held.

jonasschnelli replied:
I decided to hold the lock because it would be possible that during iteration a addWallet (or later removeWallet) happens. But right,.. it results in holding the lock for probably a long time.

I first though doing a copy of the wallets vector would be good, but the CWallet instance where the copy point to might be gone.

Maybe we keep this for now because the flushAllWallets happens only on shutdown.
Or what would be a better solution?

luke-jr commented:
#8694 is already step-by-step, so this mostly just forces rebasing it... IMO better to do refactoring after #8694 gets merged. :)

jonasschnelli commented:
Rebased and fixed code formatting (@MarcoFalkes micro-nit).
The diff is relatively large, but it would finally eliminate the global `pwalletMain` in favor of a more flexible handler that would be "ready" for multiwallet.

MarcoFalke commented:
Needs rebase

jonasschnelli commented:
Rebased (now also covers `importmulti`).

ryanofsky commented:
Seems like currently the first multiwallet commit is #8775, then #8694 comes next, and maybe this will be rebased on top of those at a later point?

jonasschnelli commented:
Closing. Since #8775, this is to far away from being useful. Maybe a different similar approach can be worked out.
